### DISCUSSION TITLE
Module not found: Can't resolve 'net'

### DISCUSSION THREAD

**Discussion by ziyadev** [vercel/next.js#62466] [Help] [answered]
### Summary

im trying to use redis on middleware but every time it tries to connect it giving me this :

```
./node_modules/.pnpm/@redis+client@1.5.14/node_modules/@redis/client/dist/lib/client/socket.js:16:12
Module not found: Can't resolve 'net'
  14 | Object.defineProperty(exports, "__esModule", { value: true });
  15 | const events_1 = require("events");
> 16 | const net = require("net");
     |             ^^^^^^^^^^^^^^
  17 | const tls = require("tls");
  18 | const errors_1 = require("../errors");
  19 | const utils_1 = require("../utils");

https://nextjs.org/docs/messages/module-not-found
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Middleware doesn't run on Node.js, it is a special version of the V8 engine, with a different set of APIs.

In this case `net` is not part of that API, https://vercel.com/docs/functions/edge-middleware/edge-runtime ~

I guess, you need to find a Redis library compatible with Edge environments.  

Check out this blog post, I think they use the upstash client to reach redis though: https://medium.com/geekculture/how-to-use-redis-with-vercel-edge-4baea510cc71
